--- col/col.c.orig	2021-04-09 02:24:12.000000000 +0200
+++ col/col.c	2022-07-28 21:56:16.961148758 +0200
@@ -47,11 +47,9 @@ static char sccsid[] = "@(#)col.c	8.5 (B
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD$");
 
-#include <sys/capsicum.h>
-
-#include <capsicum_helpers.h>
 #include <err.h>
 #include <errno.h>
+#include <limits.h>
 #include <locale.h>
 #include <stdio.h>
 #include <stdlib.h>
@@ -61,6 +59,8 @@ __FBSDID("$FreeBSD$");
 #include <wchar.h>
 #include <wctype.h>
 
+#include "compat.h"
+
 #define	BS	'\b'		/* backspace */
 #define	TAB	'\t'		/* tab */
 #define	SPACE	' '		/* space */
@@ -141,12 +141,6 @@ main(int argc, char **argv)
 
 	(void)setlocale(LC_CTYPE, "");
 
-	if (caph_limit_stdio() == -1)
-		err(1, "unable to limit stdio");
-
-	if (caph_enter() < 0)
-		err(1, "unable to enter capability mode");
-
 	max_bufd_lines = 256;
 	compress_spaces = 1;		/* compress spaces into tabs */
 	while ((opt = getopt(argc, argv, "bfhl:px")) != -1)
