--- sed/process.c.orig	2022-07-28 22:16:02.712444515 +0200
+++ sed/process.c	2022-07-28 22:16:42.516578319 +0200
@@ -669,6 +669,10 @@ regexec_e(regex_t *preg, const char *str
 	size_t start, size_t stop)
 {
 	int eval;
+#ifndef REG_STARTEND
+	char *buf;
+	size_t slen;
+#endif
 
 	if (preg == NULL) {
 		if (defpreg == NULL)
@@ -677,11 +681,26 @@ regexec_e(regex_t *preg, const char *str
 		defpreg = preg;
 
 	/* Set anchors */
+#ifndef REG_STARTEND
+	buf = malloc(stop - start + 1);
+	if (!buf)
+		errx(1, "out of memory");
+	slen = stop - start;
+	memcpy(buf, string + start, slen);
+	buf[slen] = '\0';
+	eval = regexec(defpreg, buf, nomatch ? 0 : maxnsub + 1, match, eflags);
+	for (size_t i = 0; i <= (nomatch ? 0 : maxnsub); ++i) {
+		match[i].rm_so += start;
+		match[i].rm_eo += start;
+	}
+	free(buf);
+#else
 	match[0].rm_so = start;
 	match[0].rm_eo = stop;
 
 	eval = regexec(defpreg, string,
 	    nomatch ? 0 : maxnsub + 1, match, eflags | REG_STARTEND);
+#endif
 	switch(eval) {
 	case 0:
 		return (1);
