--- cmp/special.c.orig	2022-07-28 23:02:08.880833499 +0200
+++ cmp/special.c	2022-07-28 23:02:23.367414761 +0200
@@ -40,7 +40,6 @@ __FBSDID("$FreeBSD$");
 
 #include <sys/types.h>
 
-#include <capsicum_helpers.h>
 #include <err.h>
 #include <stdlib.h>
 #include <stdio.h>
@@ -56,13 +55,6 @@ c_special(int fd1, const char *file1, of
 	FILE *fp1, *fp2;
 	int dfound;
 
-	if (caph_limit_stream(fd1, CAPH_READ) < 0)
-		err(ERR_EXIT, "caph_limit_stream(%s)", file1);
-	if (caph_limit_stream(fd2, CAPH_READ) < 0)
-		err(ERR_EXIT, "caph_limit_stream(%s)", file2);
-	if (caph_enter() < 0)
-		err(ERR_EXIT, "unable to enter capability mode");
-
 	if ((fp1 = fdopen(fd1, "r")) == NULL)
 		err(ERR_EXIT, "%s", file1);
 	if ((fp2 = fdopen(fd2, "r")) == NULL)
