--- sh/jobs.h.orig	2022-07-31 23:59:23.340660196 +0200
+++ sh/jobs.h	2022-07-31 23:59:25.648434185 +0200
@@ -55,7 +55,7 @@ void setjobctl(int);
 void showjobs(int, int);
 struct job *makejob(union node *, int);
 pid_t forkshell(struct job *, union node *, int);
-pid_t vforkexecshell(struct job *, char **, char **, const char *, int, int []);
+pid_t vforkexecshell(struct job *, char **, char **, const char *, int, int [2]);
 int waitforjob(struct job *, int *);
 int stoppedjobs(void);
 int backgndpidset(void);
