--- sh/trap.c.orig	2022-07-27 21:59:35.246239209 +0200
+++ sh/trap.c	2022-07-28 06:27:28.052029669 +0200
@@ -59,7 +59,7 @@ __FBSDID("$FreeBSD$");
 #include "mystring.h"
 #include "builtins.h"
 #include "myhistedit.h"
-
+#include "compat.h"
 
 /*
  * Sigmode records the current value of the signal handlers for the various
@@ -111,10 +111,11 @@ sigstring_to_signum(char *sig)
 
 		if (strncasecmp(sig, "SIG", 3) == 0)
 			sig += 3;
-		for (n = 1; n < sys_nsig; n++)
-			if (sys_signame[n] &&
-			    strcasecmp(sys_signame[n], sig) == 0)
+		for (n = 1; n < NSIG; n++) {
+			const char *sign = signum_to_signame(n);
+			if (sign && strcasecmp(sign, sig) == 0)
 				return (n);
+		}
 	}
 	return (-1);
 }
@@ -129,16 +130,17 @@ printsignals(void)
 	int n, outlen;
 
 	outlen = 0;
-	for (n = 1; n < sys_nsig; n++) {
-		if (sys_signame[n]) {
-			out1fmt("%s", sys_signame[n]);
-			outlen += strlen(sys_signame[n]);
+	for (n = 1; n < NSIG; n++) {
+		const char *sign = signum_to_signame(n);
+		if (sign) {
+			out1fmt("%s", sign);
+			outlen += strlen(sign);
 		} else {
 			out1fmt("%d", n);
 			outlen += 3;	/* good enough */
 		}
 		++outlen;
-		if (outlen > 71 || n == sys_nsig - 1) {
+		if (outlen > 71 || n == NSIG - 1) {
 			out1str("\n");
 			outlen = 0;
 		} else {
@@ -152,7 +154,7 @@ printsignals(void)
  * The trap builtin.
  */
 int
-trapcmd(int argc __unused, char **argv)
+trapcmd(int argc __attribute__((unused)), char **argv)
 {
 	char *action;
 	int signo;
@@ -169,14 +171,15 @@ trapcmd(int argc __unused, char **argv)
 	argv = argptr;
 
 	if (*argv == NULL) {
-		for (signo = 0 ; signo < sys_nsig ; signo++) {
+		for (signo = 0 ; signo < NSIG ; signo++) {
 			if (signo < NSIG && trap[signo] != NULL) {
+				const char *sign = signum_to_signame(signo);
 				out1str("trap -- ");
 				out1qstr(trap[signo]);
 				if (signo == 0) {
 					out1str(" EXIT\n");
-				} else if (sys_signame[signo]) {
-					out1fmt(" %s\n", sys_signame[signo]);
+				} else if (sign) {
+					out1fmt(" %s\n", sign);
 				} else {
 					out1fmt(" %d\n", signo);
 				}
@@ -509,7 +512,7 @@ exitshell_savedstatus(void)
 {
 	struct jmploc loc1, loc2;
 	char *p;
-	int sig = 0;
+	volatile int sig = 0;
 	sigset_t sigs;
 
 	if (!exiting) {
