--- m4/look.c.orig	2022-07-28 21:31:02.653653016 +0200
+++ m4/look.c	2022-07-28 21:45:13.470277493 +0200
@@ -70,20 +70,20 @@ struct ohash macros;
 
 /* Support routines for hash tables.  */
 void *
-hash_calloc(size_t n, size_t s, void *u __unused)
+hash_calloc(size_t n, size_t s, void *u UNUSED)
 {
 	void *storage = xcalloc(n, s, "hash alloc");
 	return storage;
 }
 
 void
-hash_free(void *p, void *u __unused)
+hash_free(void *p, void *u UNUSED)
 {
 	free(p);
 }
 
 void *
-element_alloc(size_t s, void *u __unused)
+element_alloc(size_t s, void *u UNUSED)
 {
 	return xalloc(s, "element alloc");
 }
