--- getopt/getopt.c.orig	2022-07-28 09:59:54.851387024 +0200
+++ getopt/getopt.c	2022-07-28 10:00:11.669739926 +0200
@@ -6,7 +6,6 @@ __FBSDID("$FreeBSD$");
  * into the public domain and is thus not subject to any copyright.
  */
 
-#include <capsicum_helpers.h>
 #include <err.h>
 #include <errno.h>
 #include <stdio.h>
@@ -19,9 +18,6 @@ main(int argc, char *argv[])
 	int c;
 	int status = 0;
 
-	if (caph_limit_stdio() < 0 || caph_enter() < 0)
-		err(1, "capsicum");
-
 	optind = 2;	/* Past the program name and the option letters. */
 	while ((c = getopt(argc, argv, argv[1])) != -1)
 		switch (c) {
