--- chroot/chroot.c.orig	2022-07-30 23:43:13.282640962 +0200
+++ chroot/chroot.c	2022-07-30 23:34:00.125038283 +0200
@@ -44,7 +44,6 @@ static char sccsid[] = "@(#)chroot.c	8.1
 __FBSDID("$FreeBSD$");
 
 #include <sys/types.h>
-#include <sys/procctl.h>
 
 #include <ctype.h>
 #include <err.h>
@@ -69,15 +68,13 @@ main(int argc, char *argv[])
 	const char	*shell;
 	gid_t		gid, *gidlist;
 	uid_t		uid;
-	int		arg, ch, error, gids;
+	int			ch, gids;
 	long		ngroups_max;
-	bool		nonprivileged;
 
 	gid = 0;
 	uid = 0;
 	user = group = grouplist = NULL;
-	nonprivileged = false;
-	while ((ch = getopt(argc, argv, "G:g:u:n")) != -1) {
+	while ((ch = getopt(argc, argv, "G:g:u:")) != -1) {
 		switch(ch) {
 		case 'u':
 			user = optarg;
@@ -94,9 +91,6 @@ main(int argc, char *argv[])
 			if (*grouplist == '\0')
 				usage();
 			break;
-		case 'n':
-			nonprivileged = true;
-			break;
 		case '?':
 		default:
 			usage();
@@ -160,13 +154,6 @@ main(int argc, char *argv[])
 		}
 	}
 
-	if (nonprivileged) {
-		arg = PROC_NO_NEW_PRIVS_ENABLE;
-		error = procctl(P_PID, getpid(), PROC_NO_NEW_PRIVS_CTL, &arg);
-		if (error != 0)
-			err(1, "procctl");
-	}
-
 	if (chdir(argv[0]) == -1 || chroot(".") == -1)
 		err(1, "%s", argv[0]);
 
@@ -193,6 +180,6 @@ static void
 usage(void)
 {
 	(void)fprintf(stderr, "usage: chroot [-g group] [-G group,group,...] "
-	    "[-u user] [-n] newroot [command]\n");
+	    "[-u user] newroot [command]\n");
 	exit(1);
 }
