--- cmp/cmp.c.orig	2022-07-28 23:00:23.875219210 +0200
+++ cmp/cmp.c	2022-07-28 23:00:58.315744261 +0200
@@ -47,7 +47,6 @@ __FBSDID("$FreeBSD$");
 #include <sys/types.h>
 #include <sys/stat.h>
 
-#include <capsicum_helpers.h>
 #include <err.h>
 #include <errno.h>
 #include <fcntl.h>
@@ -113,12 +112,6 @@ main(int argc, char *argv[])
 	if (argc < 2 || argc > 4)
 		usage();
 
-	/* Don't limit rights on stdin since it may be one of the inputs. */
-	if (caph_limit_stream(STDOUT_FILENO, CAPH_WRITE | CAPH_IGNORE_EBADF))
-		err(ERR_EXIT, "unable to limit rights on stdout");
-	if (caph_limit_stream(STDERR_FILENO, CAPH_WRITE | CAPH_IGNORE_EBADF))
-		err(ERR_EXIT, "unable to limit rights on stderr");
-
 	/* Backward compatibility -- handle "-" meaning stdin. */
 	special = false;
 	if (strcmp(file1 = argv[0], "-") == 0) {
@@ -166,9 +159,6 @@ main(int argc, char *argv[])
 			exit(ERR_EXIT);
 	}
 
-	/* FD rights are limited in c_special() and c_regular(). */
-	caph_cache_catpages();
-
 	if (!special) {
 		if (fstat(fd1, &sb1)) {
 			if (!sflag)
