--- sh/alias.c.orig	2022-07-27 21:59:35.210242734 +0200
+++ sh/alias.c	2022-07-27 22:08:29.337571044 +0200
@@ -199,7 +199,7 @@ printaliases(void)
 }
 
 int
-aliascmd(int argc __unused, char **argv __unused)
+aliascmd(int argc __attribute__((unused)), char **argv __attribute__((unused)))
 {
 	char *n, *v;
 	int ret = 0;
@@ -228,7 +228,7 @@ aliascmd(int argc __unused, char **argv
 }
 
 int
-unaliascmd(int argc __unused, char **argv __unused)
+unaliascmd(int argc __attribute__((unused)), char **argv __attribute__((unused)))
 {
 	int i;
 
