--- sh/jobs.c.orig	2022-07-27 21:59:35.226241167 +0200
+++ sh/jobs.c	2022-07-27 22:08:29.345570261 +0200
@@ -246,7 +246,7 @@ setjobctl(int on)
 
 #if JOBS
 int
-fgcmd(int argc __unused, char **argv __unused)
+fgcmd(int argc __attribute__((unused)), char **argv __attribute__((unused)))
 {
 	struct job *jp;
 	pid_t pgrp;
@@ -271,7 +271,7 @@ fgcmd(int argc __unused, char **argv __u
 
 
 int
-bgcmd(int argc __unused, char **argv __unused)
+bgcmd(int argc __attribute__((unused)), char **argv __attribute__((unused)))
 {
 	struct job *jp;
 
@@ -314,7 +314,7 @@ restartjob(struct job *jp)
 
 
 int
-jobscmd(int argc __unused, char *argv[] __unused)
+jobscmd(int argc __attribute__((unused)), char *argv[] __attribute__((unused)))
 {
 	char *id;
 	int ch, mode;
@@ -537,7 +537,7 @@ freejob(struct job *jp)
 
 
 int
-waitcmd(int argc __unused, char **argv __unused)
+waitcmd(int argc __attribute__((unused)), char **argv __attribute__((unused)))
 {
 	struct job *job;
 	int retval;
@@ -615,7 +615,7 @@ waitcmdloop(struct job *job)
 
 
 int
-jobidcmd(int argc __unused, char **argv __unused)
+jobidcmd(int argc __attribute__((unused)), char **argv __attribute__((unused)))
 {
 	struct job *jp;
 	int i;
@@ -747,7 +747,7 @@ killjob(const char *name, int sig)
  */
 
 struct job *
-makejob(union node *node __unused, int nprocs)
+makejob(union node *node __attribute__((unused)), int nprocs)
 {
 	int i;
 	struct job *jp;
@@ -1129,7 +1129,7 @@ waitforjob(struct job *jp, int *signaled
 
 
 static void
-dummy_handler(int sig __unused)
+dummy_handler(int sig __attribute__((unused)))
 {
 }
 
