--- sh/parser.c.orig	2022-05-12 06:53:55.000000000 +0200
+++ sh/parser.c	2022-07-31 23:36:50.421650694 +0200
@@ -65,6 +65,7 @@ __FBSDID("$FreeBSD$");
 #ifndef NO_HISTORY
 #include "myhistedit.h"
 #endif
+#include "compat.h"
 
 /*
  * Shell command parser.
@@ -592,6 +593,9 @@ command(void)
 	case TRP:
 		if (!redir)
 			synexpect(-1);
+		tokpushback++;
+		n1 = simplecmd(rpp, redir);
+		return n1;
 	case TWORD:
 		tokpushback++;
 		n1 = simplecmd(rpp, redir);
@@ -1092,7 +1096,6 @@ parsebackq(char *out, struct nodelist **
 	struct jmploc jmploc;
 	struct jmploc *const savehandler = handler;
 	size_t savelen;
-	int saveprompt;
 	const int bq_startlinno = plinno;
 	char *volatile ostr = NULL;
 	struct parsefile *const savetopfile = getcurrentfile();
@@ -1180,18 +1183,16 @@ parsebackq(char *out, struct nodelist **
 	(*nlpp)->next = NULL;
 
 	if (oldstyle) {
-		saveprompt = doprompt;
+		int saveprompt = doprompt;
 		doprompt = 0;
-	}
-
-	n = list(0);
-
-	if (oldstyle) {
+		n = list(0);
 		if (peektoken() != TEOF)
 			synexpect(-1);
 		doprompt = saveprompt;
-	} else
+	} else {
+		n = list(0);
 		consumetoken(TRP);
+	}
 
 	(*nlpp)->n = n;
         if (oldstyle) {
@@ -1240,12 +1241,13 @@ static char *
 readcstyleesc(char *out)
 {
 	int c, vc, i, n;
-	unsigned int v;
+	unsigned int v = 0;
 
 	c = pgetc();
 	switch (c) {
 	case '\0':
 		synerror("Unterminated quoted string");
+		return NULL; /* unreachable */
 	case '\n':
 		plinno++;
 		if (doprompt)
@@ -1989,7 +1991,7 @@ getpwlogin(void)
 static void
 getusername(char *name, size_t namelen)
 {
-	static char cached_name[MAXLOGNAME];
+	static char cached_name[LOGIN_NAME_MAX];
 	struct passwd *pw;
 	uid_t euid;
 
@@ -2024,7 +2026,7 @@ getusername(char *name, size_t namelen)
  *    should be added here.
  */
 char *
-getprompt(void *unused __unused)
+getprompt(void *unused __attribute__((unused)))
 {
 	static char ps[PROMPTLEN];
 	const char *fmt;
