--- echo/echo.c.orig	2022-07-30 22:33:43.607259803 +0200
+++ echo/echo.c	2022-07-30 22:33:38.991711825 +0200
@@ -47,7 +47,6 @@ __FBSDID("$FreeBSD$");
 #include <sys/uio.h>
 
 #include <assert.h>
-#include <capsicum_helpers.h>
 #include <err.h>
 #include <errno.h>
 #include <limits.h>
@@ -64,9 +63,6 @@ main(int argc, char *argv[])
 	char space[] = " ";
 	char newline[] = "\n";
 
-	if (caph_limit_stdio() < 0 || caph_enter() < 0)
-		err(1, "capsicum");
-
 	/* This utility may NOT do getopt(3) option parsing. */
 	if (*++argv && !strcmp(*argv, "-n")) {
 		++argv;
