--- gzip/unxz.c.orig	2022-07-28 22:37:00.741012925 +0200
+++ gzip/unxz.c	2022-07-28 22:44:57.062326563 +0200
@@ -38,6 +38,7 @@ __FBSDID("$FreeBSD$");
 #include <stdio.h>
 #include <unistd.h>
 #include <lzma.h>
+#include <sys/stat.h>
 
 static off_t
 unxz(int i, int o, char *pre, size_t prelen, off_t *bytes_in)
@@ -82,7 +83,7 @@ unxz(int i, int o, char *pre, size_t pre
 			switch (strm.avail_in) {
 			case (size_t)-1:
 				maybe_err("read failed");
-				/*NOTREACHED*/
+				break; /*NOTREACHED*/
 			case 0:
 				action = LZMA_FINISH;
 				break;
