--- stty/stty.c.orig	2021-04-09 02:24:01.000000000 +0200
+++ stty/stty.c	2022-07-28 21:56:32.067669307 +0200
@@ -51,9 +51,12 @@ __FBSDID("$FreeBSD$");
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
+#include <errno.h>
+#include <termios.h>
 
 #include "stty.h"
 #include "extern.h"
+#include "compat.h"
 
 int
 main(int argc, char *argv[])
@@ -129,10 +132,11 @@ args:	argc -= optind;
 
 		if (isdigit(**argv)) {
 			speed_t speed;
-
-			speed = strtonum(*argv, 0, UINT_MAX, &errstr);
+			unsigned int baud;
+			baud = (unsigned int)strtonum(*argv, 0, UINT_MAX, &errstr);
 			if (errstr)
 				err(1, "speed");
+			speed = get_speed(baud);
 			cfsetospeed(&i.t, speed);
 			cfsetispeed(&i.t, speed);
 			i.set = 1;
