--- sed/main.c.orig	2022-05-12 06:54:04.000000000 +0200
+++ sed/main.c	2022-07-31 22:56:30.295426706 +0200
@@ -67,6 +67,9 @@ static const char sccsid[] = "@(#)main.c
 
 #include "defs.h"
 #include "extern.h"
+#include "compat.h"
+
+extern const char *__progname;
 
 /*
  * Linked list of units (strings and files) to be compiled
@@ -213,7 +216,7 @@ usage(void)
 	(void)fprintf(stderr,
 	    "usage: %s script [-Ealnru] [-i extension] [file ...]\n"
 	    "\t%s [-Ealnu] [-i extension] [-e script] ... [-f script_file]"
-	    " ... [file ...]\n", getprogname(), getprogname());
+	    " ... [file ...]\n", __progname, __progname);
 	exit(1);
 }
 
@@ -262,7 +265,7 @@ again:
 			state = ST_STRING;
 			goto again;
 		default:
-			__unreachable();
+			abort(); /* unreachable */
 		}
 	case ST_FILE:
 		if ((p = fgets(buf, n, f)) != NULL) {
