--- sdiff/sdiff.c.orig	2022-07-28 23:06:49.549515334 +0200
+++ sdiff/sdiff.c	2022-07-30 21:02:29.326860305 +0200
@@ -28,6 +28,7 @@ __FBSDID("$FreeBSD$");
 #include <unistd.h>
 
 #include "extern.h"
+#include "compat.h"
 
 static char diff_path[] = "/usr/bin/diff";
 
@@ -260,16 +261,19 @@ main(int argc, char **argv)
 		case 'E':
 		case 'i':
 		case 't':
-		case 'W':
-			diffargv[1]  = realloc(diffargv[1], sizeof(char) * strlen(diffargv[1]) + 2);
+		case 'W': {
+			int alen = strlen(diffargv[1]);
+			diffargv[1] = realloc(diffargv[1], sizeof(char) * alen + 2);
 			/*
 			 * In diff, the 'W' option is 'w' and the 'w' is 'W'.
 			 */
 			if (ch == 'W')
-				sprintf(diffargv[1], "%sw", diffargv[1]);
+				diffargv[1][alen] = 'w';
 			else
-				sprintf(diffargv[1], "%s%c", diffargv[1], ch);
+				diffargv[1][alen] = ch;
+			diffargv[1][alen + 1] = '\0';
 			break;
+		}
 		case 'H':
 			diffargv[diffargc++] = speed_lf;
 			break;
