--- diff/diffdir.c.orig	2022-05-12 06:54:04.000000000 +0200
+++ diff/diffdir.c	2022-07-31 23:22:43.524635209 +0200
@@ -35,6 +35,7 @@ __FBSDID("$FreeBSD$");
 #include <limits.h>
 
 #include "diff.h"
+#include "compat.h"
 
 static int selectfile(const struct dirent *);
 static void diffit(struct dirent *, char *, size_t, char *, size_t, int);
@@ -57,7 +58,8 @@ diffdir(char *p1, char *p2, int flags)
 
 	dirlen1 = strlcpy(path1, *p1 ? p1 : ".", sizeof(path1));
 	if (dirlen1 >= sizeof(path1) - 1) {
-		warnc(ENAMETOOLONG, "%s", p1);
+		errno = ENAMETOOLONG;
+		warn("%s", p1);
 		status |= 2;
 		return;
 	}
@@ -67,7 +69,8 @@ diffdir(char *p1, char *p2, int flags)
 	}
 	dirlen2 = strlcpy(path2, *p2 ? p2 : ".", sizeof(path2));
 	if (dirlen2 >= sizeof(path2) - 1) {
-		warnc(ENAMETOOLONG, "%s", p2);
+		errno = ENAMETOOLONG;
+		warn("%s", p2);
 		status |= 2;
 		return;
 	}
