--- diff/diffreg.c.orig	2022-05-12 06:54:04.000000000 +0200
+++ diff/diffreg.c	2022-07-31 23:23:17.625295575 +0200
@@ -69,10 +69,8 @@
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD$");
 
-#include <sys/capsicum.h>
 #include <sys/stat.h>
 
-#include <capsicum_helpers.h>
 #include <ctype.h>
 #include <err.h>
 #include <errno.h>
@@ -85,10 +83,13 @@ __FBSDID("$FreeBSD$");
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <unistd.h>
+#include <limits.h>
+#include <time.h>
 
-#include "pr.h"
 #include "diff.h"
 #include "xmalloc.h"
+#include "compat.h"
 
 /*
  * diff - compare two files.
@@ -259,8 +260,6 @@ diffreg(char *file1, char *file2, int fl
 {
 	FILE *f1, *f2;
 	int i, rval;
-	struct pr *pr = NULL;
-	cap_rights_t rights_ro;
 
 	f1 = f2 = NULL;
 	rval = D_SAME;
@@ -339,29 +338,7 @@ diffreg(char *file1, char *file2, int fl
 		goto closem;
 	}
 
-	if (lflag)
-		pr = start_pr(file1, file2);
-
-	if (capsicum) {
-		cap_rights_init(&rights_ro, CAP_READ, CAP_FSTAT, CAP_SEEK);
-		if (caph_rights_limit(fileno(f1), &rights_ro) < 0)
-			err(2, "unable to limit rights on: %s", file1);
-		if (caph_rights_limit(fileno(f2), &rights_ro) < 0)
-			err(2, "unable to limit rights on: %s", file2);
-		if (fileno(f1) == STDIN_FILENO || fileno(f2) == STDIN_FILENO) {
-			/* stdin has already been limited */
-			if (caph_limit_stderr() == -1)
-				err(2, "unable to limit stderr");
-			if (caph_limit_stdout() == -1)
-				err(2, "unable to limit stdout");
-		} else if (caph_limit_stdio() == -1)
-				err(2, "unable to limit stdio");
-
-		caph_cache_catpages();
-		caph_cache_tzdata();
-		if (caph_enter() < 0)
-			err(2, "unable to enter capability mode");
-	}
+	(void)capsicum;
 
 	switch (files_differ(f1, f2, flags)) {
 	case 0:
@@ -424,8 +401,6 @@ diffreg(char *file1, char *file2, int fl
 	output(file1, f1, file2, f2, flags);
 
 closem:
-	if (pr != NULL)
-		stop_pr(pr);
 	if (anychange) {
 		status |= 1;
 		if (rval == D_SAME)
@@ -499,7 +474,7 @@ opentemp(const char *f)
 }
 
 char *
-splice(char *dir, char *path)
+dsplice(char *dir, char *path)
 {
 	char *tail, *buf;
 	size_t dirlen;
@@ -1376,6 +1351,7 @@ readhash(FILE *f, int flags, unsigned *h
 		case '\0':
 			if ((flags & D_FORCEASCII) == 0)
 				return (RH_BINARY);
+			/* FALLTHROUGH */
 		case '\r':
 			if (flags & D_STRIPCR) {
 				t = getc(f);
@@ -1433,7 +1409,7 @@ asciifile(FILE *f)
 static char *
 match_function(const long *f, int pos, FILE *fp)
 {
-	unsigned char buf[FUNCTION_CONTEXT_SIZE];
+	char buf[FUNCTION_CONTEXT_SIZE];
 	size_t nc;
 	int last = lastline;
 	const char *state = NULL;
@@ -1652,41 +1628,18 @@ dump_unified_vec(FILE *f1, FILE *f2, int
 static void
 print_header(const char *file1, const char *file2)
 {
-	const char *time_format;
-	char buf1[256];
-	char buf2[256];
-	char end1[10];
-	char end2[10];
-	struct tm tm1, tm2, *tm_ptr1, *tm_ptr2;
-	int nsec1 = stb1.st_mtim.tv_nsec;
-	int nsec2 = stb2.st_mtim.tv_nsec;
-
-	time_format = "%Y-%m-%d %H:%M:%S";
-
-	if (cflag)
-		time_format = "%c";
-	tm_ptr1 = localtime_r(&stb1.st_mtime, &tm1);
-	tm_ptr2 = localtime_r(&stb2.st_mtime, &tm2);
-	strftime(buf1, 256, time_format, tm_ptr1);
-	strftime(buf2, 256, time_format, tm_ptr2);
-	if (!cflag) {
-		strftime(end1, 10, "%z", tm_ptr1);
-		strftime(end2, 10, "%z", tm_ptr2);
-		sprintf(buf1, "%s.%.9d %s", buf1, nsec1, end1);
-		sprintf(buf2, "%s.%.9d %s", buf2, nsec2, end2);
-	}
 	if (label[0] != NULL)
 		printf("%s %s\n", diff_format == D_CONTEXT ? "***" : "---",
 		    label[0]);
 	else
-		printf("%s %s\t%s\n", diff_format == D_CONTEXT ? "***" : "---",
-		    file1, buf1);
+		printf("%s %s\t%s", diff_format == D_CONTEXT ? "***" : "---",
+		    file1, ctime(&stb1.st_mtime));
 	if (label[1] != NULL)
 		printf("%s %s\n", diff_format == D_CONTEXT ? "---" : "+++",
 		    label[1]);
 	else
-		printf("%s %s\t%s\n", diff_format == D_CONTEXT ? "---" : "+++",
-		    file2, buf2);
+		printf("%s %s\t%s", diff_format == D_CONTEXT ? "---" : "+++",
+		    file2, ctime(&stb2.st_mtime));
 }
 
 /*
