--- diff/diffdir.c.orig	2022-07-28 23:14:47.598579760 +0200
+++ diff/diffdir.c	2022-07-28 23:15:19.687437151 +0200
@@ -57,7 +57,8 @@ diffdir(char *p1, char *p2, int flags)
 
 	dirlen1 = strlcpy(path1, *p1 ? p1 : ".", sizeof(path1));
 	if (dirlen1 >= sizeof(path1) - 1) {
-		warnc(ENAMETOOLONG, "%s", p1);
+		errno = ENAMETOOLONG;
+		warn("%s", p1);
 		status |= 2;
 		return;
 	}
@@ -67,7 +68,8 @@ diffdir(char *p1, char *p2, int flags)
 	}
 	dirlen2 = strlcpy(path2, *p2 ? p2 : ".", sizeof(path2));
 	if (dirlen2 >= sizeof(path2) - 1) {
-		warnc(ENAMETOOLONG, "%s", p2);
+		errno = ENAMETOOLONG;
+		warn("%s", p2);
 		status |= 2;
 		return;
 	}
