--- patch/pch.c.orig	2021-04-09 02:24:12.000000000 +0200
+++ patch/pch.c	2022-07-29 00:27:54.931818006 +0200
@@ -43,6 +43,7 @@
 #include "util.h"
 #include "pch.h"
 #include "pathnames.h"
+#include "compat.h"
 
 /* Patch (diff listing) abstract type. */
 
@@ -150,13 +151,22 @@ static void
 grow_hunkmax(void)
 {
 	int new_hunkmax = hunkmax * 2;
+	void *old_line, *old_len, *old_char;
 
 	if (p_line == NULL || p_len == NULL || p_char == NULL)
 		fatal("Internal memory allocation error\n");
 
-	p_line = reallocf(p_line, new_hunkmax * sizeof(char *));
-	p_len = reallocf(p_len, new_hunkmax * sizeof(unsigned short));
-	p_char = reallocf(p_char, new_hunkmax * sizeof(char));
+	old_line = p_line;
+	old_len = p_len;
+	old_char = p_char;
+
+	p_line = realloc(p_line, new_hunkmax * sizeof(char *));
+	p_len = realloc(p_len, new_hunkmax * sizeof(unsigned short));
+	p_char = realloc(p_char, new_hunkmax * sizeof(char));
+
+	if (!p_line) free(old_line);
+	if (!p_len) free(old_len);
+	if (!p_char) free(old_char);
 
 	if (p_line != NULL && p_len != NULL && p_char != NULL) {
 		hunkmax = new_hunkmax;
@@ -1215,7 +1225,7 @@ size_t
 pgets(bool do_indent)
 {
 	char *line;
-	size_t len;
+	size_t len = 0;
 	int indent = 0, skipped = 0;
 
 	line = fgetln(pfp, &len);
