--- patch/pch.c.orig	2022-07-31 22:24:24.595818598 +0200
+++ patch/pch.c	2022-08-02 23:58:45.764185114 +0200
@@ -43,6 +43,7 @@
 #include "util.h"
 #include "pch.h"
 #include "pathnames.h"
+#include "compat.h"
 
 /* Patch (diff listing) abstract type. */
 
@@ -150,13 +151,22 @@ static void
 grow_hunkmax(void)
 {
 	int new_hunkmax = hunkmax * 2;
+	void *old_line, *old_len, *old_char;
 
 	if (p_line == NULL || p_len == NULL || p_char == NULL)
 		fatal("Internal memory allocation error\n");
 
-	p_line = reallocf(p_line, new_hunkmax * sizeof(char *));
-	p_len = reallocf(p_len, new_hunkmax * sizeof(unsigned short));
-	p_char = reallocf(p_char, new_hunkmax * sizeof(char));
+	old_line = p_line;
+	old_len = p_len;
+	old_char = p_char;
+
+	p_line = realloc(p_line, new_hunkmax * sizeof(char *));
+	p_len = realloc(p_len, new_hunkmax * sizeof(unsigned short));
+	p_char = realloc(p_char, new_hunkmax * sizeof(char));
+
+	if (!p_line) free(old_line);
+	if (!p_len) free(old_len);
+	if (!p_char) free(old_char);
 
 	if (p_line != NULL && p_len != NULL && p_char != NULL) {
 		hunkmax = new_hunkmax;
@@ -1214,14 +1224,15 @@ hunk_done:
 size_t
 pgets(bool do_indent)
 {
-	char *line;
-	size_t len;
+	char *line = NULL;
+	ssize_t len = 0;
+	size_t buflen = 0;
 	int indent = 0, skipped = 0;
 
-	line = fgetln(pfp, &len);
-	if (line != NULL) {
-		if (len + 1 > buf_size) {
-			while (len + 1 > buf_size)
+	if ((len = getline(&line, &buflen, pfp)) >= 0) {
+		char *linep = line;
+		if ((size_t)(len + 1) > buf_size) {
+			while ((size_t)(len + 1) > buf_size)
 				buf_size *= 2;
 			free(buf);
 			buf = malloc(buf_size);
@@ -1240,8 +1251,10 @@ pgets(bool do_indent)
 		}
 		memcpy(buf, line, len - skipped);
 		buf[len - skipped] = '\0';
+		line = linep;
 	}
-	return len;
+	free(line);
+	return (len > 0) ? len : 0;
 }
 
 
