--- compat/merge.c.orig	2022-05-12 06:53:59.000000000 +0200
+++ compat/merge.c	2022-07-31 00:38:29.901557053 +0200
@@ -38,6 +38,8 @@ static char sccsid[] = "@(#)merge.c	8.2
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD$");
 
+#define __is_aligned(x, align) (((uintptr_t)x & ((align) - 1)) == 0)
+
 /*
  * Hybrid exponential search/linear search merge sort with hybrid
  * natural/pairwise first pass.  Requires about .3% more comparisons
@@ -52,11 +54,15 @@ __FBSDID("$FreeBSD$");
  * (The default is pairwise merging.)
  */
 
+#include <sys/types.h>
 #include <sys/param.h>
 
 #include <errno.h>
 #include <stdlib.h>
 #include <string.h>
+#include <stdint.h>
+
+#include "compat.h"
 
 #ifdef I_AM_MERGESORT_B
 #include "block_abi.h"
