--- wall/ttymsg.c.orig	2022-07-28 20:13:16.371875970 +0200
+++ wall/ttymsg.c	2022-07-28 20:18:17.757894551 +0200
@@ -37,6 +37,7 @@ __FBSDID("$FreeBSD$");
 static const char sccsid[] = "@(#)ttymsg.c	8.2 (Berkeley) 11/16/93";
 #endif
 
+#include <sys/param.h>
 #include <sys/types.h>
 #include <sys/uio.h>
 #include <dirent.h>
@@ -68,6 +69,7 @@ ttymsg(struct iovec *iov, int iovcnt, co
 	static char errbuf[1024];
 	char *p;
 	int forked;
+	sigset_t set;
 
 	forked = 0;
 	if (iovcnt > (int)(sizeof(localiov) / sizeof(localiov[0])))
@@ -143,7 +145,8 @@ ttymsg(struct iovec *iov, int iovcnt, co
 			/* wait at most tmout seconds */
 			(void) signal(SIGALRM, SIG_DFL);
 			(void) signal(SIGTERM, SIG_DFL); /* XXX */
-			(void) sigsetmask(0);
+			sigemptyset(&set);
+			sigprocmask(SIG_SETMASK, &set, 0);
 			(void) alarm((u_int)tmout);
 			(void) fcntl(fd, F_SETFL, 0);	/* clear O_NONBLOCK */
 			continue;
