--- wall/wall.c.orig	2022-05-12 06:54:04.000000000 +0200
+++ wall/wall.c	2022-07-31 23:25:24.012917856 +0200
@@ -49,6 +49,7 @@ static const char sccsid[] = "@(#)wall.c
  */
 
 #include <sys/param.h>
+#include <sys/types.h>
 #include <sys/stat.h>
 #include <sys/uio.h>
 
@@ -162,7 +163,7 @@ main(int argc, char *argv[])
 					ingroup = 1;
 				else if ((grp = getgrgid(g->gid)) != NULL) {
 					for (np = grp->gr_mem; *np; np++) {
-						if (strcmp(*np, utmp->ut_user) == 0) {
+						if (strncmp(*np, utmp->ut_user, sizeof(utmp->ut_user)) == 0) {
 							ingroup = 1;
 							break;
 						}
