--- tail/reverse.c.orig	2021-04-09 02:24:12.000000000 +0200
+++ tail/reverse.c	2022-07-30 20:59:44.275024587 +0200
@@ -55,11 +55,16 @@ __FBSDID("$FreeBSD$");
 #include <string.h>
 #include <unistd.h>
 
-#include <libcasper.h>
-#include <casper/cap_fileargs.h>
-
 #include "extern.h"
 
+/* some versions of queue.h don't come with this */
+#ifndef TAILQ_FOREACH_FROM_SAFE
+#define TAILQ_FOREACH_FROM_SAFE(var, head, field, tvar)                 \
+	for ((var) = ((var) ? (var) : TAILQ_FIRST((head)));             \
+	    (var) && ((tvar) = TAILQ_NEXT((var), field), 1);            \
+	    (var) = (tvar))
+#endif
+
 static void r_buf(FILE *, const char *);
 static void r_reg(FILE *, const char *, enum STYLE, off_t, struct stat *);
 
